server/utils/passport/local-login.js

```diff
@@ -19,64 +20,47 @@ export default (req, res, next, user) => {
     done(null, user)
   })
   //If the user exists, serialize the user to the session
-  req.login(user, function (err) {
-    const { appDb } = req.app.locals
-    const uid = user.uid
+  req.login(user, async function (err) {
+    if (err) return res.redirect(`${routes.login}?e=${err}`)
 
-    appDb.collection(collections.configs).findOne(
-      {
-        owner: uid,
-      },
-      function (err, configObj) {
-        if (err) {
-          console.error(err.message)
-        }
-        if (!configObj) {
-          const defaultConfig = {
-            owner: uid,
-            config: defaultUserConfig,
-            name: 'default',
-            type: 'matrix',
-            readers: [uid],
-            created: new Date().toUTCString(),
-          }
-          appDb.collection(collections.configs).insertOne(defaultConfig)
-        }
-        appDb.collection(collections.users).findOneAndUpdate(
-          { uid: user.uid },
-          {
-            $set: {
-              last_logon: Date.now(),
-            },
-          },
-          {
-            projection: {
-              _id: 0,
-              uid: 1,
-              display_name: 1,
-              acl: 1,
-              role: 1,
-              icon: 1,
-              mail: 1,
-              access: 1,
-            },
-            returnOriginal: false,
-            upsert: true,
-          },
-          function (err, userinfo) {
-            if (err) {
-              console.error(err.message)
-            }
-            req.session.role = userinfo.value.role
-            req.session.display_name = userinfo.value.display_name
-            req.session.mail = userinfo.value.mail
-            req.session.celery_tasks = []
-            req.session.icon = userinfo.value.icon
-            req.session.userAccess = userinfo.value.access
-            return res.redirect(`${basePath}/`)
-          }
-        )
+    try {
+      const { appDb } = req.app.locals
+      const uid = user.uid
+      const config = await ConfigModel.findConfig(appDb, { owner: uid })
+
+      if (!config) {
+        const configAttributes = { owner: uid, readers: [uid] }
+
+        await ConfigModel.save(appDb, configAttributes)
       }
-    )
+
+      const userInfo = await UserModel.findUserAndUpdate(
+        appDb,
+        {
+          uid: user.uid,
+        },
+        {
+          last_logon: Date.now(),
+        }
+      )
+      const { role, display_name, mail, icon, access, account_expires } =
+        userInfo.value
+      const today = moment()
+      const accountExpirationToMoment = moment(account_expires)
+      const isAccountExpired = accountExpirationToMoment.isBefore(today)
+
+      if (isAccountExpired) return res.redirect(`${routes.login}?e=forbidden`)
+
+      req.session.role = role
+      req.session.display_name = display_name
+      req.session.mail = mail
+      req.session.celery_tasks = []
+      req.session.icon = icon
+      req.session.userAccess = access
```

Comment:
I've made this comment a few times but I don't think it's been addressed. Do we need to set these session variables or can we just set the user UID in the session and get the user's attributes from querying for the user with that UID?

Reply:
Can you help me understand what you mean?

Reply:
This does answer my question. I think there is opportunity for refactor here but agree it can happen at a later date.

In theory, we should only need to access the `uid` from the session. We can use that `uid` to get the user for that session and then get these other attributes from the user. This would also mean updating parts of the UI that reference these session variables to use the values from the user object instead.